plugins/tiff/src/tiffpresentation.cc

```diff
@@ -195,28 +195,39 @@ bool TiffPresentation::load(const std::string& fileName_)
     printf("This bitmap has size %d*%d, aspect ratio %.1f*%.1f\n",
            width, height, 1/resolutionX, 1/resolutionY);
 
+    LayerSpec ls;
     if (spp == 4 && bps == 8)
     {
       ls.push_back(OperationsCMYK32::create());
+      properties[PIPETTE_PROPERTY_NAME] = "";
+      this->pipetteLayer = boost::dynamic_pointer_cast<PipetteLayerOperations>(ls[0]);
```

Comment:
No need for the `this->` (this isn't java :wink: )

By changing the order around, you can avoid the dynamic cast:
```
pipetteLayer = OperationsCMYK32::create();
ls.push_back(pipetteLayer);
properties[PIPETTE_PROPERTY_NAME] = "";
```

... well... That would have worked if `PipetteLayerOperations` actually inherited from `LayerOperations`. Perhaps you should reconsider :wink:

Reply:
This is an interesting suggestion that I just tried out. Essentially this would at the current stage introduce a diamond hierarchy (`PipetteCommonOperations` would inherit from both `PipetteLayerOperations` and `CommonOperations`; Both `PipetteLayerOperations` and `CommonOperations` would inherit from `LayerOperations`). While this can work, i believe there are some problems which need to be fixed and i could not really identify the best approach (the Compiler is mostly complaining about some methods not being implemented in `OperationsCMYKx`/ `Operations24bpp`, probably because it is confused as to where it is implemented; The methods are `initializeCairo(), draw()` & `drawState()`):

1. Cut the link between `PipetteCommonOperations` and `CommonOperations`. This would mean that all methods defined in `CommonOperations` would need to be re-implemented in `PipetteCommonOperations`. This creates quite a lot of code duplication...
2. Re-declare the methods from `CommonOperations` in `PipetteCommonOperations` and delegate the call to `CommonOperations`. This would be my own preference, however does not feel that clean...
3. Restructure the whole hierarchy. Essentially, `CommonOperations` needs to inherit from `LayerOperations`, and we also want `PipetteLayerOperations`to inherit from `LayerOperations`. `PipetteCommonOperations` also needs to inherit from `PipetteLayerOperations`. The only link which I think could be cut would be from `PipetteCommonOperations` to `CommonOperations`. However, that would introduce code duplication...
4. Continue using the current solution with dynamic casts. They are not a great solution but perhaps we can reconsider these given the above problem.

These are the ideas that i currently have. Maybe you have a better solution for the above problem

Reply:
Oops... That sounds more involved than I anticipated. I think of all of the above, I favor alternative 4 for simplicity.

There's one more variation I'd like you to consider
```
auto cmykOperations = OperationsCMYK32::create();
pipetteLayer = cmykOperations;
ls.push_back(cmykOperations);
properties[PIPETTE_PROPERTY_NAME] = "";
```
This is slightly longer than we had before, but has two upcasts, which are done implicitly, and can be compile-time-checked. As opposed to one cross-cast, which theoretically could fail at runtime.

Reply:
`OperationsCMYK32::create()` does return a `LayerOperations::Ptr` right now. How about changing this to `PipetteCommonOperations::Ptr` (and of course adding a typedef for the `PipetteCommonOperations`)?

Reply:
Oeps... Yes, that would make sense, then.